crates/pera-core/src/dwallet_mpc/mpc_outputs_verifier.rs

```diff
@@ -0,0 +1,165 @@
+//! by checking if a validators with quorum of stake voted for it.
+//! Any validator that voted for a different output is considered malicious.
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use pera_types::base_types::{AuthorityName, ObjectID};
+use pera_types::committee::StakeUnit;
+use pera_types::messages_dwallet_mpc::SessionInfo;
+use std::collections::{HashMap, HashSet};
+
+/// A struct to verify the DWallet MPC outputs.
+/// It stores all the outputs received for each session,
+/// and decides whether an output is valid
+/// by checking if a validators with quorum of stake voted for it.
+pub struct DWalletMPCOutputsVerifier {
+    /// The outputs received for each MPC session.
+    pub mpc_sessions_outputs: HashMap<ObjectID, SessionOutputsData>,
+    /// A mapping between an authority name to its stake.
+    /// todo(zeev): can we use the data from the manager?
+    pub weighted_parties: HashMap<AuthorityName, StakeUnit>,
+    /// The quorum threshold of the chain.
+    pub quorum_threshold: StakeUnit,
+    pub completed_locking_next_committee: bool,
+    voted_to_lock_committee: HashSet<AuthorityName>,
+}
+
+/// The data needed to manage the outputs of an MPC session.
+#[derive(Clone)]
+pub struct SessionOutputsData {
+    // todo(zeev): cleanup.
+    /// Maps session's output to the authorities that voted for it.
+    pub session_output_to_voting_authorities:
+        HashMap<(Vec<u8>, SessionInfo), HashSet<AuthorityName>>,
+    /// Needed to make sure an authority does not send two outputs for the same session.
+    pub authorities_that_sent_output: HashSet<AuthorityName>,
+}
+
+/// The result of verifying an incoming output for an MPC session.
+/// We need to differentiate between a duplicate and a malicious output,
+/// as the output can be sent twice by honest parties.
+#[derive(PartialOrd, PartialEq)]
+pub enum OutputResult {
+    Valid,
+    Malicious,
+    /// We need more votes to decide if the output is valid or not.
+    NotEnoughVotes,
+}
+
+pub struct OutputVerificationResult {
+    pub result: OutputResult,
+    pub malicious_actors: Vec<AuthorityName>,
+}
+
+impl DWalletMPCOutputsVerifier {
+    pub fn new(epoch_store: &AuthorityPerEpochStore) -> Self {
+        DWalletMPCOutputsVerifier {
+            quorum_threshold: epoch_store.committee().quorum_threshold(),
+            mpc_sessions_outputs: HashMap::new(),
+            weighted_parties: epoch_store
+                .committee()
+                .voting_rights
+                .clone()
+                .into_iter()
+                .collect(),
+            completed_locking_next_committee: false,
+            voted_to_lock_committee: HashSet::new(),
+        }
+    }
+
+    /// Returns true if the `lock_next_epoch_committee` system TX should get called, a.k.a. a quorum of validators voted for it,
+    /// and false otherwise.
+    pub(crate) fn should_lock_committee(&mut self, authority_name: AuthorityName) -> bool {
```

Comment:
The logic of this function also _inserts a new vote_ for the locking of the committee, but neither its name nor its description reflect that.

```suggestion
    pub(crate) fn add_vote_and_maybe_lock_committee(&mut self, authority_name: AuthorityName) -> bool {
```